### PR TITLE
add logprobs endpoint with caching past key values

### PR DESCRIPTION
**Pull request by yangkevin2**
Hi, 

I'd like to be able to modify the logits before sampling at each step of generation, so I implemented a logprobs endpoint which returns logprobs based on top-p and top-k and also caches past_key_values. To avoid mixing work items with the existing completions endpoint, I made a separate queue for the new endpoint. The past_key_values cache removes old items after 10 minutes or after it gets too full. 

I was hoping to integrate the new endpoint more with the existing completions endpoint which calls huggingface's generate() method, but it seems like this might not be possible at the moment due to generate() not supporting returning past key values pending this PR https://github.com/huggingface/transformers/pull/17574. 

Also, the new endpoint does everything with batch size 1 to avoid needing to mix past_key_values from different calls, but then pads up to batch size MAX_BS=16 anyway since that's what the model interface seems to expect. Not sure if this would be a problem down the line. 

Please let me know if there's something implemented in a problematic way and I can try to fix it. Would really like to be able to run controlled generation approaches with these larger models.

### CODE REVIEW

File: examples/opt_serving/interactive_hosted.py [hunk 1]
```diff
@@ -235,6 +313,63 @@ def completions(engine=None):
     return OAIResponse(results).__dict__()
 
 
+@app.route("/logprobs", methods=["POST"])
```
Comment (zhuohan123) [inline examples/opt_serving/interactive_hosted.py:316]: Can we have a flag to turn on/off this API when starting the interactive host? And also, can we set a password or security key so that we can only allow private access to this API?

File: examples/opt_serving/interactive_hosted.py [hunk 2]
```diff
@@ -105,13 +116,80 @@ def batching_loop(timeout=TIMEOUT_MS, max_tokens=MAX_BATCH_TOKENS, max_bs=MAX_BS
...
+                    for work_item, gen in zip(logprobs_batch, generations):
+                        work_item.return_queue.put((work_item.uid, gen))
+
+                    logprobs_batch.clear()
+        
+        # clear old keys in PAST_CACHE which haven't been used recently
+        existing_cache_ids = list(PAST_CACHE.keys())
+        for cache_id in existing_cache_ids:
+            if time.time() - PAST_CACHE[cache_id][0] > PAST_CACHE_TIMEOUT:
+                del PAST_CACHE[cache_id]
```
Comment (zhuohan123) [inline examples/opt_serving/interactive_hosted.py:191]: If I understand correctly, PAST_CACHE will be a huge cache that occupies a large portion of GPU memory. If we only use the TIMEOUT mechanism, will it make our original service OOM due to the cache in GPU?

**Review by zhuohan123 — CHANGES_REQUESTED** [review]
Left some review comments. @zhisbug please also review this PR since it affects our main service.

File: examples/opt_serving/generator.py [hunk 3]
```diff
@@ -137,6 +137,50 @@ def load_model(self):
...
+        for new_input in new_inputs:
+            ori_len = len(new_input)
+            if len(new_input) < max_len:
+                new_input.extend([pad_value for _ in range(max_len - ori_len)])
+        # pad to bs = MAX_BS
+        if bs < MAX_BS:
+            new_inputs.extend([[pad_value for _ in range(max_len)] for _ in range(MAX_BS - bs)])
+        return new_inputs
+    
+    def __call__(
```
[the remark this answers is not in the document]
Comment (yangkevin2) [reply examples/opt_serving/generator.py:158]: I used `__call__` because it's essentially just wrapping the OPT model's forward function, which would normally be accessed by just doing e.g. `model(**args)`. I could rename it to `forward` instead and explicitly call `.forward()` if that would be preferred? What do you think?

File: examples/opt_serving/generator.py [hunk 4]
```diff
@@ -137,6 +137,50 @@ def load_model(self):
...
+        self,
+        inputs,
+        cache_ids,
+        pasts=None,
+    ):
+        assert len(inputs) == 1, "enforcing batch size 1 for now to avoid messy logic with caching past key values"
```
[the remark this answers is not in the document]
Comment (yangkevin2) [reply examples/opt_serving/generator.py:164]: I plan to call it frequently, but only one request (batch size 1) at a time since the requests are sequential in nature, so I don't think the batching would make things faster for my use case. So I didn't implement it for now. Is that fine for the current PR?

File: examples/opt_serving/interactive_hosted.py [hunk 5]
```diff
@@ -105,13 +116,80 @@ def batching_loop(timeout=TIMEOUT_MS, max_tokens=MAX_BATCH_TOKENS, max_bs=MAX_BS
...
+                for work_item, gen in zip(generate_batch, generations):
                     work_item.return_queue.put((work_item.uid, gen))
 
-                batch.clear()
-            else:
-                # back to the loop
-                continue
+                generate_batch.clear()
+
+        try: # logprobs endpoint
```
[the remark this answers is not in the document]
Comment (yangkevin2) [reply examples/opt_serving/interactive_hosted.py:124]: Sure, will refactor

File: examples/opt_serving/interactive_hosted.py (hunk 1 above)
Comment (yangkevin2) [reply examples/opt_serving/interactive_hosted.py:316]: Yeah I can definitely look into adding a flag and have it off by default or something. I'm not sure what's the best practice for the password/security key though, is there some method you have in mind?

File: examples/opt_serving/interactive_hosted.py (hunk 2 above)
Comment (yangkevin2) [reply examples/opt_serving/interactive_hosted.py:191]: I also wrote a mechanism to remove the oldest item once it exceeds a specified maximum size, regardless of timeout, based on `PAST_CACHE_SIZE_LIMIT`. So the cache should be limited to some maximum size. 

Additionally, it might be possible to cache only the relevant keys instead of the full padded batch of size 16 by manipulating the past_key_values, but I wasn't sure if constantly reconstructing the padded keys on GPU would introduce a lot of overhead so I didn't do it for now. 

I'm not sure how much of a limitation the memory would be on your server; happy to modify the numbers/implementation based on what you think the right tradeoff is.

**Comment by zhisbug** [comment]
@yangkevin2 Thanks for the great work -- I'll review by tonight

File: examples/opt_serving/interactive_hosted.py [hunk 6]
```diff
@@ -49,15 +56,26 @@ def batching_loop(timeout=TIMEOUT_MS, max_tokens=MAX_BATCH_TOKENS, max_bs=MAX_BS
...
-    batch = []
+    generate_batch = []
+    logprobs_batch = []
     while True:
-        try:
-            # for now, we only have 1 worker, so can always index to shard 0
-            target_queue = BATCH_QUEUE.queue_shards[0].get_largest_queue()
-            if not target_queue:
-                continue
+        generate_batch, made_model_call = generate_loop(generate_batch, timeout=timeout, max_tokens=max_tokens, max_bs=max_bs)
```
Comment (zhisbug) [inline examples/opt_serving/interactive_hosted.py:68]: I suppose this means if the external users are submitting requests, the returned `made_model_call` of their requests will be True. So your endpoint won't run?

But if at any point, you are submitting requests while any other users are not submitting requests (hence the old batching queue is empty), your requests will go through into the `if not made_model_call` branch, which will then go to the function you added in `generator.py`? 

Please correct if some of my understanding is something wrong.

File: examples/opt_serving/interactive_hosted.py [hunk 7]
```diff
@@ -247,6 +397,12 @@ def index():
     parser.add_argument("--model", type=str, default="alpa/opt-125m")
     parser.add_argument("--path", type=str, default="/home/ubuntu/opt_weights/")
     parser.add_argument("--port", type=int, default=DEFAULT_PORT)
+    parser.add_argument("--serve-logprobs", action="store_true", default=False)
     args = parser.parse_args()
 
+    if args.serve_logprobs:
+        logprobs.provide_automatic_options = False
+        logprobs.methods = ['POST']
```
Comment (zhisbug) [inline examples/opt_serving/interactive_hosted.py:405]: How do you plan to query the endpoint? Do you have client code? Would also be good to share.

Also, we can setup this on AWS and use your client code to do a few tests.

After things are tested stable, we can discuss a time window (say, some weekend 1 - 5am) when your program can run with full capacity

File: examples/opt_serving/interactive_hosted.py [hunk 8]
```diff
@@ -67,51 +85,126 @@ def batching_loop(timeout=TIMEOUT_MS, max_tokens=MAX_BATCH_TOKENS, max_bs=MAX_BS
...
+                    if key in ro:
+                        request_object[key] = ro[key]
                 # do the actual generations
                 request_object["seed"] = random.randint(1, 20000)
-                generations = generator.generate(**request_object)
+                outputs = generator(request_object['inputs'], cache_ids, pasts=PAST_CACHE)
+                generations = []
+                for cache_id, output in zip(cache_ids, outputs):
+                    PAST_CACHE[cache_id] = (time.time(), output.past_key_values) # add to or update the cache with newly computed values
+                    if len(PAST_CACHE) > PAST_CACHE_SIZE_LIMIT: # delete oldest key in cache if cache too big
```
Comment (zhisbug) [inline examples/opt_serving/interactive_hosted.py:185]: What does this cache mean? is it the attention cache or something else?

**Review by zhisbug — CHANGES_REQUESTED** [review]

**Comment by zhisbug** [comment]
@yangkevin2 please take a look at my comments! thanks you!

**Comment by yangkevin2** [comment]
Thanks for the detailed feedback! I'm out traveling this weekend but will make changes accordingly on Monday.

File: examples/opt_serving/generator.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (merrymercy) [reply examples/opt_serving/generator.py:158]: SGTM

File: examples/opt_serving/generator.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (merrymercy) [reply examples/opt_serving/generator.py:164]: If you only submit one request at a time, the speed will be very slow!
1. You have to pay the network overhead for each request. But with batching, you can amortize the overhead
2. The latency of batch-size =1 and batch-size = 16 is the same. By batching, you can get 16x better throughput.

It is fine for the current PR because it is a separate endpoint. Later, if you find the generation speed too slow, you may apply batching

File: examples/opt_serving/generator.py [hunk 9]
```diff
@@ -137,6 +137,50 @@ def load_model(self):
...
+            dataset=self.build_dataset_for_inference(tokens, lengths),
+            max_tokens=None,
+            max_sentences=MAX_BS,
+            max_positions=None,
+            ignore_invalid_inputs=False,
+        ).next_epoch_itr(shuffle=False)
+        assert len(batches) == 1, "we have enforce inner bs = 1."
+        outputs = []
+        for idx, (batch, cache_id) in enumerate(zip(batches, cache_ids)):
+            input_ids = move_to_cuda(batch['src_tokens'])[:, :-1]
```
Comment (merrymercy) [inline examples/opt_serving/generator.py:178]: We change the default torch device to CPU (while alpa still runs on GPU) to avoid the memory conflicts between torch and alpa.

Please update accordingly
https://github.com/alpa-projects/alpa/blob/6085a0877a999d60cfc4409e25ef51cd96703eea/examples/opt_serving/generator.py#L171-L172

**Review by merrymercy — COMMENTED** [review]
It looks good to me generally. During last few days, we made some big improvements, such as beam search (https://github.com/alpa-projects/alpa/pull/615), faster generation (https://github.com/alpa-projects/alpa/pull/661), and code clean-up (https://github.com/alpa-projects/alpa/pull/667). Please rebase!

File: examples/opt_serving/interactive_hosted.py [hunk 10]
```diff
@@ -20,7 +23,11 @@
 from opt_serving.service.constants import MAX_SEQ_LEN, MAX_BATCH_TOKENS, DEFAULT_PORT, TIMEOUT_MS, MAX_BS
 
 app = Flask(__name__)
-BATCH_QUEUE = PriorityQueueRingShard()
+GENERATE_BATCH_QUEUE = PriorityQueueRingShard()
+LOGPROBS_BATCH_QUEUE = PriorityQueueRingShard()
+PAST_CACHE = defaultdict(lambda: (0, None))
```
[the remark this answers is not in the document]
Comment (yangkevin2) [reply examples/opt_serving/interactive_hosted.py]: Done!

File: examples/opt_serving/interactive_hosted.py [hunk 11]
```diff
@@ -67,51 +85,126 @@ def batching_loop(timeout=TIMEOUT_MS, max_tokens=MAX_BATCH_TOKENS, max_bs=MAX_BS
...
+                ro = work_item.data
+                request_object["inputs"].append(ro["input"])
+                request_object["min_tokens"].append(ro.get("min_tokens", 0))
+                request_object["max_tokens"].append(
+                    ro.get("max_tokens", MAX_SEQ_LEN))
+                cache_ids.append(ro.get("cache_id", None))
+                # assumption: everyone has the same remaining args
+                for key in [
+                        "temperature",
+                        "top_p",
```
[the remark this answers is not in the document]
Comment (yangkevin2) [reply examples/opt_serving/interactive_hosted.py:172]: Yeah, I envisioned possibly wanting to have both so I implemented both for this endpoint (but didn't touch the completions endpoint)

File: examples/opt_serving/generator.py [hunk 12]
```diff
@@ -137,6 +137,50 @@ def load_model(self):
...
+            attention_mask = self.model_wrapper._prepare_attention_mask_for_generation(input_ids, pad_token_id=self.model_wrapper.config.pad_token_id, eos_token_id=self.model_wrapper.config.eos_token_id)
+            model_inputs = self.model_wrapper.prepare_inputs_for_generation(input_ids, past=pasts[cache_id][1] if pasts is not None else None, attention_mask=attention_mask)
```
[the remark this answers is not in the document]
Comment (yangkevin2) [reply examples/opt_serving/generator.py:106]: I did notice that the current batching logic was being taken from the first sentence and being applied to everything else, but ignored this issue because in my current implementation of the new endpoint it's just forcing batch size 1. So I think it should be ok?

File: examples/opt_serving/interactive_hosted.py (hunk 6 above)
Comment (yangkevin2) [reply examples/opt_serving/interactive_hosted.py:72]: Yeah, Zhuohan suggested that I prioritize the existing endpoint, so I implemented this logic to prioritize requests to completions and only run my endpoint if the queue is empty.

File: examples/opt_serving/interactive_hosted.py (hunk 8 above)
Comment (yangkevin2) [reply examples/opt_serving/interactive_hosted.py]: It's a cache for the past_key_values, yeah.

File: examples/opt_serving/interactive_hosted.py (hunk 7 above)
Comment (yangkevin2) [reply examples/opt_serving/interactive_hosted.py:437]: Here's some simple example code I've been using to test:

```
import requests
import json

from scipy.special import softmax
import numpy as np
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained('facebook/opt-125m')

url = "http://localhost:9054/logprobs"
headers = {'Content-type': 'application/json', 'Accept': 'text/plain'}
data = {
    "prompt":[int(x) for x in tokenizer.encode("mary had a little")],
    "max_tokens":"64",
    "temperature":"0.7",
    "top_p":"0.5",
    "top_k":"100",
}

r = requests.post(url, data=json.dumps(data), headers=headers)
# import pdb; pdb.set_trace()

for i in range(20):
    output = r.json()
    data['cache_id'] = output['cache_id']
    distribution = np.array([-1e8 for _ in range(tokenizer.vocab_size)])
    print(len(output['logprobs']))
    for idx, logprob in zip(output['indices'], output['logprobs']):
        distribution[idx] = logprob
    distribution = softmax(distribution)
    token = np.random.choice(np.arange(len(distribution)), p=distribution)
    data['prompt'].append(int(token))
    print(tokenizer.decode(data['prompt']))
    r = requests.post(url, data=json.dumps(data), headers=headers)

import pdb; pdb.set_trace()
```

After things are tested + stable, I was hoping to be able to query the endpoint at any time, and just give priority to the completions endpoint when there's traffic (as I implemented with the `made_model_call` logic). Would that be possible?

File: examples/opt_serving/generator.py (hunk 9 above)
Comment (yangkevin2) [reply examples/opt_serving/generator.py]: Updated!

**Comment by yangkevin2** [comment]
Hi, I've rebased and (I think) addressed all the existing comments, minus @zhisbug 's comments about the incorrect generation params with batching since I'm currently not batching with the new endpoint anyway.

Thinking about it more carefully, I may actually want to batch though, so I'll work on that this week and see if it's feasible (might be tricky with having to also cache past_key_values; if necessary I could force the user to pass one entire batch at a time or something, and advance the batch all together, to avoid mixing batches). 

I also noticed that when testing on AWS, it took a while longer to generate sequentially one token at a time using the new endpoint compared to just using the normal completions endpoint, even when I ran the client code from the same machine. Not sure if there's still network latency with python's requests module even when running from the same machine or if there's some other bottleneck; will dig around a bit more. 

Also, an alternative implementation that doesn't require so many network calls could be to use the `logits_warper` functionality in huggingface's generate (https://huggingface.co/docs/transformers/main_classes/text_generation#transformers.generation_utils.GenerationMixin.sample.logits_warper). This might actually be cleaner code-wise since it'd just be an extra argument to the completions endpoint, but might require being able to pass in arbitrary functions as a logits_warper and running them on the server, so might need some extra security? Curious if you guys would prefer going this route or the current one.

**Comment by merrymercy** [comment]
I port your code and did some clean-up and fixes in #669 
The major fix is the batching logic.
1. Your current implementation of the priority is not correct. We should use `receive_item` instead of `made_model_call` to implement priority
2. The slow generation speed also comes from the batching logic. The default code has a `timeout=2000ms` argument, so it always waits for 2 seconds to wait for batching opportunities before actually responding to a request. Now, I set a timeout=0 for the logprob endpoint, so it is much faster.

We can go ahead with the current approach and let you start trying things. If you find the speed too slow, I think `logits_warper` might be the best way to go. We can give you a password and only support this API with a correct password, so other people cannot run arbitrary python code. If we want to implement the `logits_warper` API, the question is "Can all your logits modification code be wrapped in a stateless python function?"

**Comment by merrymercy** [comment]
The logprob endpoint and a test script are added (https://github.com/alpa-projects/alpa/blob/main/examples/opt_serving/test_logprobs.py).
I tried to set it up on our server. The completion API works, but I got `403 Forbidden` error for the logprobs API. I guess we need to update some firewall settings, @zhisbug will figure this out.

**Comment by yangkevin2** [comment]
Great, thank you for fixing the priority + timeout! Will keep the `logits_warper` in mind but I think it might not be necessary for now. 

Anyway, please just let me know when the main OPT-175B server is ready to serve logprobs requests, and/or if anything comes up that I need to fix.
